add FAQ page

dcalvin opened the pull request:
Signed-off-by: Calvin Weng <wenghao@pingcap.com>

Added FAQ page. @Hoverbear  On further thought and investigation, I think FAQ should be in parallel with other top-level menus like Reference and Tasks, which are the practice of many other documentations like TiDB and CRDB.

dcalvin commented:
I think I will try if I can pull more FAQ content from the TIDB FAQ tomorrow and update this doc further.

Hoverbear commented on content/docs/3.0/faq/faq.md:
```diff
@@ -0,0 +1,151 @@
+- NUMA: no specific suggestion; for memory allocation strategy, you can use `interleave = all`
+- File system: ext4
+
+## Can Raft + multiple replicas in the TiKV architecture achieve absolute data security? Is it necessary to apply the most strict mode (`sync-log = true`) to a standalone storage?
```
What is a standalone storage?

> a standalone storage?

dcalvin replied:
A standalone TiKV

Hoverbear commented on content/docs/3.0/faq/faq.md:
```diff
@@ -0,0 +1,151 @@
+
+## Does TiDB have a InnoDB memcached plugin like MySQL which can directly use the KV interface and does not need the independent cache?
+
+TiKV supports calling the interface separately. Theoretically, you can take an instance as the cache. Because TiDB is a distributed relational database, we do not support TiKV separately.
```
```suggestion
TiKV supports calling the interface separately. Theoretically, you can take an instance as the cache.
```

TiDB isn't relevant here. :)

Hoverbear commented on content/docs/3.0/faq/faq.md:
```diff
@@ -0,0 +1,151 @@
+
+## How does Region split?
+
+Region is not divided in advance, but it follows a Region split mechanism. When the Region size exceeds the value of the `region_split_size` or `region-split-keys` parameters, split is triggered. After the split, the information is reported to PD.
```
Are the parameter names *actually* this inconsistent? (`_` vs `-`)?

dcalvin replied:
Good catch. It's region-split-size

Hoverbear commented on content/docs/3.0/faq/faq.md:
```diff
@@ -0,0 +1,151 @@
+
+## If a node is down, will the service be affected? How long?
+
+TiKV uses Raft to synchronize data among multiple replicas and guarantees the strong consistency of data. If one replica goes wrong, the other replicas can guarantee data security. The default number of replicas in each Region is 3. Based on the Raft protocol, a leader is elected in each Region, and if a single leader fails, a follower is soon elected as Region leader after a maximum of 2 * lease ...
```
I don't think this answer actually answers the question, does it? It just says "Yes your data is secure" it doesn't say that the service should remain available (potentially after a short reelection time)

dcalvin replied:
Right! @youjiali1995 @hicqu  Can you help "re-answer" this question?

BusyJay replied:
Besides, we usually say "safety" instead of "security". "security" means your data won't be compromised by hackers, "safety" means your data won't suffer from corruption or loss.

Hoverbear commented on content/docs/3.0/faq/faq.md:
```diff
@@ -0,0 +1,151 @@
+
+## How does TiKV compare to NoSQL databases like Cassandra, HBase, or MongoDB?
+
+TiKV is as scalable as NoSQL databases but features in the usability and functionality of traditional SQL databases, such as SQL syntax and consistent distributed transactions.
```
```suggestion
TiKV is scalable, as NoSQL databases like to advertise, while including features like externally consistent distributed transactions and good support for stateless query layers such as TiSpark (Spark), Titan (Redis), or TiDB (MySQL).
```

Hoverbear commented on content/docs/3.0/faq/faq.md:
```diff
@@ -0,0 +1,151 @@
+
+## How does TiKV compare to traditional relational databases like Oracle or MySQL?
+
+TiKV scores in horizontal scalability while still maintains the traditional relational database features. You can easily increase the capacity or balance the load by adding more machines.
```
```suggestion
TiKV is horizontally scalable while maintaining many traditional relational database features. You can easily increase the capacity or balance the load by adding more machines.
```

Hoverbear commented on content/docs/3.0/faq/faq.md:
```diff
@@ -0,0 +1,151 @@
+Yes. ACID semantics are guaranteed in TiKV:
+
+* Atomicity: Each transaction in TiKV is "all or nothing": if one part of the transaction fails, then the entire transaction fails, and the database state is left unchanged. TiKV guarantees atomicity in each and every situation, including power failures, errors, and crashes. 
+* Consistency: TiKV ensures that any transaction brings the database from one valid state to another. Any data written to the TiKV database must be valid according to all defined rules, including constraints, cascades, triggers, and any combination thereof. 
```
```suggestion
* Consistency: TiKV ensures that any transaction brings the database from one valid state to another. Any data written to the TiKV database must be valid according to all defined rules.
```

iirc we don't support the other things (Also TiDB doesn't support triggers, why are we saying it does here?)

Hoverbear commented on content/docs/3.0/faq/faq.md:
```diff
@@ -0,0 +1,151 @@
+
+* Atomicity: Each transaction in TiKV is "all or nothing": if one part of the transaction fails, then the entire transaction fails, and the database state is left unchanged. TiKV guarantees atomicity in each and every situation, including power failures, errors, and crashes. 
+* Consistency: TiKV ensures that any transaction brings the database from one valid state to another. Any data written to the TiKV database must be valid according to all defined rules, including constraints, cascades, triggers, and any combination thereof. 
+* Isolation: TiKV provides snapshot isolation (SI), snapshot isolation with lock (SQL statement: SELECT ... FOR UPDATE), and externally consistent reads and writes in distributed transactions. 
```
```suggestion
* Isolation: TiKV provides snapshot isolation (SI), snapshot isolation with lock, and externally consistent reads and writes in distributed transactions. 
```

Nope we don't do SQL in TiKV! :)

Hoverbear commented on content/docs/3.0/faq/faq.md:
```diff
@@ -0,0 +1,151 @@
+* Atomicity: Each transaction in TiKV is "all or nothing": if one part of the transaction fails, then the entire transaction fails, and the database state is left unchanged. TiKV guarantees atomicity in each and every situation, including power failures, errors, and crashes. 
+* Consistency: TiKV ensures that any transaction brings the database from one valid state to another. Any data written to the TiKV database must be valid according to all defined rules, including constraints, cascades, triggers, and any combination thereof. 
+* Isolation: TiKV provides snapshot isolation (SI), snapshot isolation with lock (SQL statement: SELECT ... FOR UPDATE), and externally consistent reads and writes in distributed transactions. 
+* Durability: TiKV allows a collection of machines to work as a coherent group that can survive the failures of some of its members. So in TiKV, once a transaction has been committed, it will remain so, even in the event of power loss, crashes, or errors. 
```
```suggestion
* Durability: TiKV allows a collection of machines to work as a coherent group that can survive the failures of a minority of its members. So in TiKV, once a transaction has been committed, it will remain so, even in the event of power loss, crashes, or errors. 
```

Hoverbear commented on content/docs/3.0/faq/faq.md:
```diff
@@ -0,0 +1,151 @@
+
+## What is the recommended number of replicas in the TiKV cluster? Is it better to keep the minimum number for high availability?
+
+Use 3 replicas for testing. If you increase the number of replicas, the performance declines but it is more secure. Whether to configure more replicas depends on the specific business needs.
```
This doesn't really answer the question. Maybe this is better?

```suggestion
You should never operate a TiKV cluster with under 3 nodes in a production scenario. Depending on infrastructure, workload, and resiliency needs, you may wish to increase this.
```

Eventually, this should become it's own page.

Hoverbear replied:
@dcalvin PTAL

BusyJay commented on content/docs/3.0/faq/faq.md:
```diff
@@ -0,0 +1,151 @@
+
+Yes. For now, we have a [TiKV driver for Go](https://github.com/pingcap/tidb/tree/master/store/tikv). You are welcome to develop drivers in other languages!
+
+## How does TiKV compare to traditional relational databases like Oracle or MySQL?
```
No, you should not compare tikv to oracle or mysql.

dcalvin replied:
What do you suggest? Remove this question?

BusyJay replied:
I suggest to answer it with “TiKV is a distributed transactional key-value database, while Oracle and MySQL are relational databases. TiDB may be more suitable to be compared with them."

dcalvin replied:
> TiDB may be more suitable to be compared with them.
In this case I don't think this question is worth listing here... So I suggest removing it.

dcalvin commented:
@zhangjinpeng1987 @BusyJay @disksing @Hoverbear I've addressed most of the comments and pulled some FAQs from TiDB docs. PTAL and let me know if there is anything that is not accurate or stale.

disksing commented on content/docs/3.0/faq/faq.md:
```diff
@@ -0,0 +1,166 @@
+
+Grow TiKV as your business grows. You can increase the capacity simply by adding more machines. You can run TiKV across physical, virtual, container, and cloud environments.
+
+PD([Placement Driver](https://github.com/pingcap/pd))  periodically checks replication constraints and balances the load, and it handles data movement automatically. When PD notices that the load is too high, it will rebalance data.
```
PD ([Placement Driver](https://github.com/pingcap/pd)) periodically checks replication constraints and balances the load, and it handles data movement automatically. When PD notices that the load is too high, it will rebalance data.
```suggestion
PD ([Placement Driver](https://github.com/pingcap/pd)) periodically checks replication constraints and balances the load, and it handles data movement automatically. When PD notices that the load is too high, it will rebalance data.
```

disksing commented on content/docs/3.0/faq/faq.md:
```diff
@@ -0,0 +1,166 @@
+
+- The default CF stores real data and the corresponding parameter is in `[rocksdb.defaultcf]`. The write CF stores the data version information (MVCC) and index-related data, and the corresponding parameter is in `[rocksdb.writecf]`. The lock CF stores the lock information and the system uses the default parameter.
+- The Raft RocksDB instance stores Raft logs. The default CF mainly stores Raft logs and the corresponding parameter is in `[raftdb.defaultcf]`.
+- Each CF has an individual block-cache to cache data blocks and improve RocksDB read speed. The size of block-cache is controlled by the `block-cache-size` parameter. A larger value of the parameter means more hot data can be cached and is more favorable to read operation. At the same time, it consumes more system memory.
```
I think now block cache of CFs are merged together.

dcalvin replied:
How about:

"CFs have a shared block-cache to cache data blocks and improve RocksDB read speed."

disksing replied:
LGTM

disksing commented on content/docs/3.0/faq/faq.md:
```diff
@@ -0,0 +1,166 @@
+
+## Is the Range of the Key data table divided before data access?
+
+No. It differs from the table splitting rules of MySQL. In TiKV, the table Range is dynamically split based on the size of Region.
```
I think current version does split by table by default.

dcalvin replied:
Can you suggest the text?

Hoverbear replied:
@disksing Note these docs are TiKV specific, and don't concern ourselves with TiDB.

disksing commented on content/docs/3.0/faq/faq.md:
```diff
@@ -0,0 +1,166 @@
+
+## Can Raft + multiple replicas in the TiKV architecture achieve absolute data safety? Is it necessary to apply the most strict mode (`sync-log = true`) to a standalone storage?
+
+Data is redundantly copied between TiKV nodes using the [Raft consensus algorithm](https://raft.github.io/) to ensure recoverability should a node failure occur. Only when the data has been written into more than 50% of the replicas will the application return ACK (two out of three nodes). However, theoretically, two nodes might crash. Therefore, except for scenarios with less strict requirement  ...
```
copied -> replicated?

Hoverbear commented on content/docs/3.0/faq/faq.md:
```diff
@@ -0,0 +1,166 @@
+
+## How do I run TiKV?
+
+For further information, see [Quick Start](../tasks/getting-started.md) for deploying a TiKV testing cluster, or [Deploy TiKV](../tasks/deploy/introduction.md) for deploying TiKV in production.
```
Please check these links, I think they won't work.

Hoverbear commented on content/docs/3.0/faq/faq.md:
```diff
@@ -0,0 +1,166 @@
+
+## Can I use TiKV as a key-value store?
+
+Yes. For now, we have a stable [TiKV driver for Go](https://github.com/pingcap/tidb/tree/master/store/tikv). There are also drivers for [other languages in development](../reference/clients/introduction.md).
```
```suggestion
Yes, that is what TiKV is!
```

Hoverbear commented on content/docs/3.0/faq/faq.md:
```diff
@@ -0,0 +1,166 @@
+* Strong consistency
+* Support for distributed ACID transactions
+
+## When not to use TiKV?
```
```suggestion
## When is TiKV inappropriate?
```
